nntrainer/layers/common_properties.h

```diff
@@ -1115,6 +1136,138 @@ class MoL_K : public PositiveIntegerProperty {
   using prop_tag = uint_prop_tag;             /**< property type */
 };
 
+/**
+ * @brief NumHeads property, NumHeads is number of head in multi head attention
+ *
+ */
+class NumHeads : public PositiveIntegerProperty {
+public:
+  /**
+   * @brief Construct a new NumHeads object with default value 1
+   *
+   */
+  NumHeads(unsigned int value = 1);
+  static constexpr const char *key = "num_heads"; /**< unique key to access */
+  using prop_tag = uint_prop_tag;                 /**< property type */
+};
+
+/**
+ * @brief QueryDim property, input query dim in multi head attention
+ * @details Correspond with embed_dim of torch
+ *
+ */
+class QueryDim : public PositiveIntegerProperty {
+public:
+  static constexpr const char *key = "query_dim"; /**< unique key to access */
+  using prop_tag = uint_prop_tag;                 /**< property type */
+};
+
+/**
+ * @brief KeyDim property, input key dim in multi head attention
+ * @details Correspond with kdim of torch
+ *
+ */
+class KeyDim : public PositiveIntegerProperty {
+public:
+  static constexpr const char *key = "key_dim"; /**< unique key to access */
+  using prop_tag = uint_prop_tag;               /**< property type */
+};
+
+/**
+ * @brief ValueDim property, input value dim in multi head attention
+ * @details Correspond with vdim of torch
+ *
+ */
+class ValueDim : public PositiveIntegerProperty {
+public:
+  static constexpr const char *key = "value_dim"; /**< unique key to access */
+  using prop_tag = uint_prop_tag;                 /**< property type */
+};
+
+/**
+ * @brief ProjectedKeyDim property, projected key dim in multi head attention
+ * @details Correspond with key_dim of tensorflow
+ *
+ */
+class ProjectedKeyDim : public PositiveIntegerProperty {
```

Comment:
I wonder what is the difference betwwen `KeyDim` and `ProjectedKeyDim`?

Reply:
It is not a alias and used for actual different purpose.
In multi head attention layer each input(query, key, value) will be projected by 3 different fc layer.(query_fc, key_fc, value_fc)
The KeyDim means input key width dimension which also means input size of key_fc layer.
The projectedKeyDim means projected key width dimension which also means unit size of key_fc layer.

There is 2 different algorithm in multi head attention layer of torch/tensorflow.
1. If the return_attention_weight is set tensorflow return attention weight before applying dropout whereas pytorch will return the attention weight which is already applied the dropout.
2. The true/false value behaves differently in attention mask. In tensorflow 1 indicates attention and 0 indicates no attention whereas 1 indicates no attention and 0 indicates attention in pytorch

Reply:
Like we discussted off-line, nntrainer does not need to work as other frameworks. The compatibility with other frameworks can be considered in the model loader and if it is possible to convert them in model loader level, then we don't need to implement like this. Optimizer Adam is different. It is the case which is different the algorithm itself.

Reply:
Done